### ISSUE TITLE
[ПЕРЕВІРЕНО — ПОТРІБНІ ЗМІНИ] запит 2164: зважена оцінювальна втрата

### ISSUE THREAD

**Issue by Oleksii-debug** [Oleksii-debug/12-6-ai.#2165]
SWARM_PROTOCOL: SWARM-300-V2
SWARM_CONTROL: #723
PARENT: #2113 #802 #548
TARGET_PR: #2164
TARGET_EXACT_HEAD: 0dd6c6f59bd8cdd683d32f94dbe073409d3697f8
TARGET_BASE_BRANCH: model341/20m-candidate-a-20260826
TARGET_BASE_SHA: b3a3353751fe0258a0c49bc2b26e85c5f995f08f
ROLE_NEEDED: FRESH DIFFERENT-WORKER READ-ONLY AUDITOR
PRODUCT_MUTATION: false
MERGE_AUTHORITY: false
TRAINING_AUTHORITY: false
PAID_COMPUTE_AUTHORITY: false
FINAL_TEST_ACCESS: false

AUDIT SCOPE:
1. Verify `_mean_loss` aggregates additive NLL over the count of unmasked shifted causal targets rather than arithmetic mean of per-batch means.
2. Verify unequal split vs merged batching invariance, `-100` masking, equal-count control, empty collection and all-masked rejection.
3. Confirm no trainer/model/tokenizer/corpus/checkpoint/evaluation-source authority changes.
4. Inspect numerical/type edge cases and fail closed if target counting can diverge from `causal_lm_loss` semantics.
5. Because the child PR targets canonical carrier #802 rather than default main, absence of an automatic shared workflow run is not itself PASS. Run lawful focused validation or provide an exact reproducible review path; do not fabricate CI.

SCIENTIFIC TRUTH UNCHANGED:
AUTHORIZED_OPTIMIZED_TARGET_EXPOSURE=0
TOKENIZER_FIT_AUTHORIZED=false
OPTIMIZER_UPDATES_EXECUTED_ON_REAL_TARGETS=0
TRAINING_EXECUTED=false
LEARNED_WEIGHTS_CREATED=false
FINAL_TEST_OUTCOMES_READ=false
PAID_COMPUTE_USED=false
FOREIGN_PRETRAINED_WEIGHTS=false

A passing verdict authorizes only selective intake of this two-file metric repair into #802.

**Comment by Oleksii-debug** [comment]
CHANGES_REQUIRED — fresh different-worker exact-head audit.

TARGET_PR: #2164
AUDITED_HEAD: 0dd6c6f59bd8cdd683d32f94dbe073409d3697f8
AUDIT_OWNER: #2168
PRODUCT_MUTATION: false
MERGE_AUTHORITY: false
TRAINING_AUTHORITY: false
PAID_COMPUTE_AUTHORITY: false
FINAL_TEST_ACCESS: false
AUDIT_OWNERSHIP_RELEASED: true

ARBITRATION
- OPEN issues were directly enumerated newest-first through physical empty page 16; exact audit lane matched #2168 only.
- OPEN pull requests were separately enumerated through physical empty page 7; no competing exact audit lane was found.
- Final late-bind refresh confirms PR #2164 remains OPEN/non-draft at the same exact head.

SEMANTIC RESULT
The Product change is mathematically correct. `causal_lm_loss` shifts labels by one and computes mean cross-entropy while ignoring `-100`. The repaired `_mean_loss` counts exactly those same shifted non-`-100` targets, converts each batch mean back to additive NLL by multiplying by that count, sums across batches, then divides by the total target count. Unequal partitioning therefore no longer biases the metric. Empty collection and an all-masked batch fail closed; non-finite/negative loss is rejected. The Product delta is limited to the evaluator plus focused tests and grants no corpus/training/model/tokenizer/checkpoint authority.

TERMINAL BLOCKER — mechanical lint failure
Exact-head workflow run 36027008041 is terminal FAILURE. Preflight passed, but locked x86-64 job 107726374848 failed the repository checks on:
`tests/test_s0_nll_weighting.py:12`
`def eval(self) -> "_DummyModel":`
Ruff UP037 requires the already-resolvable postponed annotation to be unquoted:
`def eval(self) -> _DummyModel:`

This is a one-line test-only repair; it does not invalidate the mathematical approach. However a failing exact-head locked check cannot receive PASS.

MINIMUM SAME-LINEAGE REPAIR
1. Remove the unnecessary quotes around `_DummyModel` in that return annotation (or apply the equivalent Ruff fix).
2. Run the existing exact-head checks on the new head.
3. Request a fresh exact-head audit. Do not widen Product scope.

VERDICT: CHANGES_REQUIRED for exact head 0dd6c6f59bd8cdd683d32f94dbe073409d3697f8.
